### PR TITLE
CIF-1153 - Generic related product component

### PR DESCRIPTION
**Pull request by cjelger**
- created new component based on product carousel
- added unit tests
- adapted carousel component and sling model to support the new component

## How Has This Been Tested?

Added unit tests, and manually tested in AEM.

## Screenshots (if appropriate):

Screenshot 2019-12-13 at 09 55 35

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/relatedproducts/RelatedProductsRetriever.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import com.adobe.cq.commerce.magento.graphql.ProductInterface;
+import com.adobe.cq.commerce.magento.graphql.ProductInterfaceQueryDefinition;
+import com.adobe.cq.commerce.magento.graphql.ProductPricesQueryDefinition;
+import com.adobe.cq.commerce.magento.graphql.ProductsQueryDefinition;
+import com.adobe.cq.commerce.magento.graphql.QueryQuery;
+
+class RelatedProductsRetriever extends AbstractProductsRetriever {
+
+    static enum RelationType {
+        RELATED_PRODUCTS, UPSELL_PRODUCTS, CROSS_SELL_PRODUCTS
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/relatedproducts/RelatedProductsRetriever.java]: If we design the enum with key & name we could use a more author-friendly name for the dialog data source.

Comment (cjelger) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/relatedproducts/RelatedProductsRetriever.java]: You're right, I'll change this to display user-friendly names.

Comment (dplaton) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/relatedproducts/RelatedProductsRetriever.java]: > If we design the enum with key & name we could use a more author-friendly name for the dialog data source.

...and be able to translate those names as well.

Comment (cjelger) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/relatedproducts/RelatedProductsRetriever.java]: I added human-readable text, and they should be translatable with the use of `ResourceBundle`.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/productcarousel/ProductCarousel.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
 public interface ProductCarousel {
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/productcarousel/ProductCarousel.java]: One more thing. I'm thinking we should rename interface now as this now more than just a carousel thing. But I'm not really creative with a more common name.

Comment (cjelger) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/productcarousel/ProductCarousel.java]: Well it's still a carousel, why change that? Only the source of the data changes ... but it's still  a carousel of products.

Comment (mhaack) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/productcarousel/ProductCarousel.java]: Well was just an idea :-) lets keep it.

**Pull request merged**